vpn-proxy/app/models.py

```diff
@@ -135,3 +153,82 @@ def save(self, *args, **kwargs):
     def delete(self, *args, **kwargs):
         self.stop()
         super(Tunnel, self).delete(*args, **kwargs)
+
+
+class Ports(models.Model):
+	src_addr = models.GenericIPAddressField(protocol='IPv4')
+	dst_addr = models.GenericIPAddressField(protocol='IPv4', validators=[check_destination_ip])
+	dst_port = models.IntegerField()
+	tunel_id = models.IntegerField()
+	loc_port = models.IntegerField(unique=True)
+	created_at = models.DateTimeField(auto_now_add=True)
+
+	@property
+	def tunnel(self):
+		return '%s%s' % (IFACE_PREFIX, self.tunel_id)
+
+	@property
+	def port(self):
+		return self.loc_port
+
+	@property
+	def rtable(self):
+		return 'rt_%s' % self.tunnel
+
+	@property
+	def destination(self):
+		return '%s:%s' % (self.dst_addr, self.dst_port)
+
+	def __str__(self):
+		return '%s at local port %s via %s -> %s:%s' % (self.src_addr, self.port, self.tunnel, self.dst_addr, self.dst_port)
+
+	def to_dict(self):
+		return {
+			'id': self.id,
+			'src_addr': self.src_addr,
+			'dst_addr': self.dst_addr,
+			'dst_port': self.dst_port,
+			'dst_pair': self.destination,
+			'tunnel_id': self.tunel_id,
+			'tunnel_name': self.tunnel,
+			'loc_port': self.loc_port,
+			'r_table': self.rtable
+		}
+
+	def forward(self, mode, **new_record):
+		modes = {
+			'add': '-A',
+			'delete': '-D'
+		}
+		for key in modes.iterkeys():
+			if key == mode:
+				job = modes[key]
+				# mangle incoming packets based on local port
+				# mangle table is traversed before nat in every chain
+				run(['sudo', 'iptables', '-t', 'mangle', job, 'PREROUTING', '-p', 'tcp',
+				     '--destination-port', str(new_record['loc_port']),
+				     '-j', 'MARK', '--set-mark', str(new_record['tunnel_id'])])
+				# DNAT incoming packets in order to force forwarding --> private host (IP, PORT)
+				run(['sudo', 'iptables', '-t', 'nat', job, 'PREROUTING', '-p', 'tcp',
+				     '--destination-port', str(new_record['loc_port']),
+				     '-j', 'DNAT', '--to-destination', str(new_record['dst_pair'])])
+				# MASQUERADE packets routed via the virtual interface
+				run(['sudo', 'iptables', '-t', 'nat', job, 'POSTROUTING', '-p', 'tcp', '-o', str(new_record['tunnel_name']),
+				     '-d', str(new_record['dst_addr']), '--destination-port', str(new_record['dst_port']),
+				     '-j', 'MASQUERADE'])
+				# point marked packets to the corresponding routing table as created during `openvpn start`
+				run(['sudo', 'ip', 'rule', str(key), 'fwmark', str(new_record['tunnel_id']), 'table', str(new_record['r_table'])])
```

Comment:
Try to make all these work in an idempotent way. No matter how many times one starts to run this function, it shouldn't add duplicate rules. See the functions in tunnels.py to see how they do that